### PR TITLE
Fix border condition in Snowbridge free consensus Updates

### PR DESCRIPTION
**Pull request by claravanstaden**
# Description

A fix for a border condition introduced with new feature https://github.com/paritytech/polkadot-sdk/pull/5201. A malicious relayer could spam the Ethereum client with sync committee updates that have already been imported for the period. This PR adds a storage item to track the last imported sync committee period, so that subsequent irrelevant updates are not free.

Original PR: https://github.com/Snowfork/polkadot-sdk/pull/172

## Integration

Downstream projects are not affected. Relayers will not be able to spam the Ethereum client with irrelevant sync committee updates for free.

## Review Notes

Adds a storage item to track the last free sync committee update period, so that duplicate imports are not free.

### CODE REVIEW

**Comment by claravanstaden** [comment]
@acatangiu may you please add the [T15-bridges](https://github.com/paritytech/polkadot-sdk/labels/T15-bridges) label. :)

**Review by acatangiu — APPROVED** [review]

**Review by vgeddes — APPROVED** [review]

**Comment by acatangiu** [comment]
@bkontur @serban300 please review - we also need to backport this fix.

**Review by alistair-singh — APPROVED** [review]

**Review by franciscoaguirre — COMMENTED** [review]
I'd add another check after this that updates with a lower number also are not free

**Comment by claravanstaden** [comment]
> I'd add another check after this that updates with a lower number also are not free

@franciscoaguirre do you mean another test check or a check in the light client?

**Comment by claravanstaden** [comment]
@franciscoaguirre [more test conditions](https://github.com/paritytech/polkadot-sdk/pull/5671/commits/af4f5ce0c237234c73c1fdb76333db25c779c58f) added.

@acatangiu @yrong @vgeddes @alistair-singh I added an interesting test case [here](https://github.com/paritytech/polkadot-sdk/pull/5671/files#diff-ea6025b3cac6719f750d1f0065ba30d70a16d386a5ce374e4a6b79299f48a3eeR748). It is an update with a sync committee that has already been imported. The update contains a newer finalized header, so the update is free. However, the sync committee merkle proof is verified, but not stored, because it is already known and stored. So technically the sync committee merkle proof computation is unnecessary, but the update is still free because it contains a new finalized header update. I think it is OK, but just highlighting it is an odd case.

**Comment by yrong** [comment]
> but the update is still free because it contains a new finalized header update

a new finalized header with slot advanced more than `FreeHeadersInterval` which can't be spammed, right?

File: bridges/snowbridge/pallets/ethereum-client/src/lib.rs [hunk 1]
```diff
@@ -179,6 +179,10 @@ pub mod pallet {
 	#[pallet::storage]
 	pub type NextSyncCommittee<T: Config> = StorageValue<_, SyncCommitteePrepared, ValueQuery>;
 
+	/// The last period where the next sync committee was updated for free.
+	#[pallet::storage]
+	pub type LatestFreeSyncCommitteeUpdatePeriod<T: Config> = StorageValue<_, u64, ValueQuery>;
```
Comment (serban300) [inline bridges/snowbridge/pallets/ethereum-client/src/lib.rs:184]: Would it make sense to rename this to `LatestSyncCommitteeUpdatePeriod` ? Looks like it's set even if the sync committee update is not free.

**Review by serban300 — APPROVED** [review]
Looks good as far as I can understand !

**Comment by claravanstaden** [comment]
> a new finalized header update with slot advanced more than `FreeHeadersInterval` which can't be spammed, right?

@yrong correct. So it is a valid, free update. The only weird thing is the sync committee will also be [verified](https://github.com/paritytech/polkadot-sdk/blob/master/bridges/snowbridge/pallets/ethereum-client/src/lib.rs#L380-L400) and [prepared](https://github.com/paritytech/polkadot-sdk/blob/master/bridges/snowbridge/pallets/ethereum-client/src/lib.rs#L446-L452) but not imported, which is unnecessary.

File: bridges/snowbridge/pallets/ethereum-client/src/lib.rs (hunk 1 above)
Comment (claravanstaden) [reply bridges/snowbridge/pallets/ethereum-client/src/lib.rs]: That's a good point! :) https://github.com/paritytech/polkadot-sdk/pull/5671/commits/6e4f99daf4d2ffeccefbbd8a4ab3f1eaabe077a7

**Comment by acatangiu** [comment]
@claravanstaden 

```
error[E0412]: cannot find type `LatestFreeSyncCommitteeUpdatePeriod` in this scope
   --> bridges/snowbridge/pallets/ethereum-client/src/lib.rs:479:6
    |
184 |     pub type LatestSyncCommitteeUpdatePeriod<T: Config> = StorageValue<_, u64, ValueQuery>;
    |     --------------------------------------------------------------------------------------- similarly named type alias `LatestSyncCommitteeUpdatePeriod` defined here
...
479 |                 <LatestFreeSyncCommitteeUpdatePeriod<T>>::set(update_finalized_period);
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: a type alias with a similar name exists: `LatestSyncCommitteeUpdatePeriod`

error[E0433]: failed to resolve: use of undeclared type `LatestFreeSyncCommitteeUpdatePeriod`
   --> bridges/snowbridge/pallets/ethereum-client/src/lib.rs:667:36
    |
667 |             let latest_free_update_period = LatestFreeSyncCommitteeUpdatePeriod::<T>::get();
    |                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |                                             |
    |                                             use of undeclared type `LatestFreeSyncCommitteeUpdatePeriod`
    |                                             help: a type alias with a similar name exists: `LatestSyncCommitteeUpdatePeriod`
```

quick find+replace to rename across the whole repo required.

**Pull request merged**